Fixes #2626 : Introduce MockSettings.mockMaker

JojOatXGME opened the pull request:
<!-- Hey,
Thanks for the contribution, this is awesome.
As you may have read, project members have somehow an opinionated view on what and how should be
Mockito, e.g. we don't want mockito to be a feature bloat.
There may be a thorough review, with feedback -> code change loop.
-->
<!--
If you have a suggestion for this template you can fix it in the .github/PULL_REQUEST_TEMPLATE.md file
-->

I implemented a prototype to fix #2626 as suggested in https://github.com/mockito/mockito/issues/2589#issuecomment-1105163928. The basics doesn't seem that difficult, but I could have missed some edge cases. I mostly skipped documentation and tests for now. Furthermore, I ignored that `ByteBuddyMockMaker` and `InlineByteBuddyMockMaker` are not exposed by the API, so we still have to adjust the API to allow setting the mock maker without accessing these classes.

Personally, I don't need this functionality right now. However, if you think it would be a worthwhile addition, I can still try to help with that topic.

This pull request allows users of Mockito to choose different MockMakers per mock. This feature could ease the migration from one MockMaker to another in a big project.

```
SomeClass subclassMock = Mockito.mock(SomeClass.class, Mockito.withSettings()
    .mockMaker(AwesomeMockMaker.class));
```

~~PS: The Gradle task `spotlessApply` failed on my system with *“java.lang.NoClassDefFoundError: Could not initialize class com.google.googlejavaformat.java.java14.Java14InputAstVisitor”*. Probably because it couldn't find `BindingPatternTree`. Not sure if it is a local problem or something with the project configuration.~~

## Checklist

 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [x] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [x] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [x] Avoid other runtime dependencies
 - [ ] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [x] The pull request follows coding style
 - [x] Mention `Fixes #<issue number>` in the description _if relevant_
 - [x] At least one commit should mention `Fixes #<issue number>` _if relevant_

JojOatXGME commented:
I was able to run `spotlessApply` from the terminal (PowerShell on Windows). I previously tried to run the task from IntelliJ which for some reason was using a development build of JDK 14. 😅

TimvdLippe commented on src/main/java/org/mockito/internal/util/MockUtil.java:
```diff
@@ -18,22 +18,55 @@
+    private static MockMaker getMockMaker(Class<? extends MockMaker> type) {
+        if (type == null) {
+            return defaultMockMaker;
+        } else {
```
Nit: remove the else so that the return is on the same level as the if

JojOatXGME replied:
Done. For me, it feels easier to read with the else, but I guess that might just be because I am used to that style.

TimvdLippe commented on src/main/java/org/mockito/internal/util/MockUtil.java:
```diff
@@ -104,10 +131,24 @@ public static boolean isMock(Object mock) {
+        return getMockHandler0(mock) != null;
+    }
+
+    private static MockHandler<?> getMockHandler0(Object mock) {
```
Can we use a more descriptive name here? What about `maybeGetMockHandler`?

JojOatXGME replied:
I renamed it to `getMockHandlerOrNull`.

TimvdLippe commented on src/test/java/org/mockito/CustomMockMakerTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.Arrays;
+
+public final class CustomMockMakerTest {
```
Can we also add support for `@Mock` and add a corresponding test as well?

Additionally, we are missing a bit of code coverage. Would it be possible to write some additional regression tests for these?

JojOatXGME replied:
I added an option for `@Mock` and a corresponding test. I did not add an option for `@Spy`, through.

JojOatXGME commented on src/main/java/org/mockito/internal/stubbing/defaultanswers/RetrieveGenericsForDefaultAnswers.java:
```diff
@@ -38,6 +38,7 @@ static Object returnTypeForMockWithCorrectGenerics(
 
         if (type != null) {
+            // TODO: Should we use the mockMaker of the mock?
             if (!MOCKITO_CORE.isTypeMockable(type)) {
```
Done.

JojOatXGME commented on src/main/java/org/mockito/internal/configuration/SpyAnnotationEngine.java:
```diff
@@ -83,6 +83,7 @@ public AutoCloseable process(Class<?> context, Object testInstance) {
     }
 
     private static Object spyInstance(Field field, Object instance) {
+        // TODO: Should we also add an option to @Spy?
```
I did not add an option for now, because it would be the first option on that annotation.

JojOatXGME commented on src/main/java/org/mockito/internal/util/MockUtil.java:
```diff
@@ -29,35 +29,26 @@
-    }
+    private static final Map<Class<? extends MockMaker>, MockMaker> mockMakers =
+            new ConcurrentHashMap<>(
+                    Collections.singletonMap(defaultMockMaker.getClass(), defaultMockMaker));
```
I removed that `WeakHashMap` because I realized that it wouldn't remove anything anyway. (The class cannot be garbage collected when we are still holding the instance as the value.) However, I think not being able to unload a class loader because we have a reference to the mock maker is an acceptable limitation. I could not think of a realistic solution. (We could dynamically generate classes in the class loader which belongs to the mock maker to hold the reference, but I think that would be overkill. 😄)

JojOatXGME commented on src/main/java/org/mockito/internal/util/reflection/FieldInitializer.java:
```diff
@@ -259,8 +259,11 @@ public int compare(Constructor<?> constructorA, Constructor<?> constructorB) {
+                            //  It looks like we are not actually creating any mocks for these
+                            //  parameters, but might only use mocks which are already created
+                            //  and stored in some fields.
+                            if (MockUtil.typeMockabilityOf(aClass, null).mockable()) {
```
Not sure how to handle this.

JojOatXGME reviewed (commented):
Here are my changes. I hope I am able to respond faster the next time. We also have to think about the API. Should the user be able to provide arbitrary mock makers (as right now in the PR), or should we make it more specific to the upcoming migration.

**Generic**

```
withSettings().mockMaker(SomeMockMaker.class)
```

In this case, we need to expose the internal mock makers somehow, or provide separate methods in the settings interface.

**Specific**

```
withSettings().useSubclassMockMaker()
```

Keeping the interface more specific to the upcoming migration would make it easier to deprecate and remove the feature in the future if it causes maintenance problems for some reason.

JojOatXGME commented on src/test/java/org/mockito/CustomMockMakerTest.java:
```diff
@@ -0,0 +1,189 @@
+import org.mockito.exceptions.verification.SmartNullPointerException;
+import org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker;
+import org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker;
+import org.mockito.internal.creation.bytebuddy.SubclassByteBuddyMockMaker;
```
> Unfortunately, we can't expose these classes in Mockito.java, since we don't know which mockmaker we should load.

Maybe we could use static methods to make the class objects available without loading or exposing the class?

```java
public final class MockMakers {
  private MockMakers() {} // Cannot be instantiated

  public Class<? extends MockMaker> inline() { return InlineByteBuddyMockMaker.class; };
  public Class<? extends MockMaker> subclass() { return SubclassByteBuddyMockMaker.class; };
  // ...
}
```

In contrast to constants, I think this should only load the classes for which you call the method. I am unsure why it is important to not load the classes, though.

TimvdLippe replied:
I wonder if we should instead use the String aliases for the mockmakers. E.g. you can set an enum here, of which the string value will correspond with the alias defined in https://github.com/mockito/mockito/blob/de63c24352d035dff77e9ef8e17481d9d85e87d9/src/main/java/org/mockito/internal/configuration/plugins/DefaultMockitoPlugins.java#L22

Then we would need to introduce an alias for the subclass mockmaker as well, such as `mockito-subclass-maker`?

JojOatXGME replied:
I guess strings or providing two separate options are our only possibility. At least I cannot think of anything else. I am not sure where you want me to use an enum. I would just use a class with string constants instead.

TimvdLippe replied:
Yeah agreed. The easiest solution here is to allow users to specify the String fully qualified name of the mockmaker. Since we don't want to expose the internal mockmaker locations as public API, for Mockito itself we would need to support the alias for the `mock-maker-inline` and `mock-maker-subclass`.

And disregard the comment regarding enums. That wouldn't work for community mockmakers, which would be a major benefit for them. If we lock the possible values to just the enum, that wouldn't work for them. Good point 👍

TimvdLippe replied:
@JojOatXGME Shall we go for the String solution including the default alias for `mock-maker-inline` and a new alias `mock-maker-subclass`?

JojOatXGME replied:
At least, I don't have a good idea for an alternative. I adjusted the implementation. However, I duplicated the code for the aliases for the time being. I was not sure how to share the code between `MockUtil.getMockMaker` and `PluginInitializer.loadImpl` without making bigger changes.

TimvdLippe commented on src/main/java/org/mockito/internal/util/MockUtil.java:
```diff
@@ -4,10 +4,14 @@
 import org.mockito.internal.configuration.plugins.Plugins;
+import org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker;
+import org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker;
+import org.mockito.internal.creation.proxy.ProxyMockMaker;
```
Since we do special loading in these mockmakers, we can't load their classes until we are certain that the user wants them. Therefore, we will need to dynamically load these classes using reflection (based on the names). In other words, similar to what `DefaultMockitoPlugins` does, where we also refer to them as fully qualified names rather than the type class.

The `getMockMaker` implementation down below would then becomes something along the lines of:

```java
        Class<? extends MockMaker> type;
        switch (mockMaker) {
            case MockMakers.INLINE:
            case MockMakers.PROXY:
            case MockMakers.SUBCLASS:
                type = defaultMockitoPlugins.getDefaultPluginClass(mockMaker);
                break;
            default:
                {
                    ClassLoader loader = Thread.currentThread().getContextClassLoader();
                    if (loader == null) {
                        loader = ClassLoader.getSystemClassLoader();
                    }
                    try {
                        type = loader.loadClass(mockMaker).asSubclass(MockMaker.class);
                    } catch (Exception e) {
                        throw new IllegalStateException(
                                "Failed to load MockMaker: " + mockMaker, e);
                    }
                    break;
                }
        }
```

(Store the `defaultMockitoPlugins` in a field as a constant and make `getDefaultPluginClass` public, it's an internal class anyways so not part of our API)

If you feel adventurous, it would be good to add an integration test to ensure this remains so. I think it would be good to copy https://github.com/mockito/mockito/blob/main/src/test/java/org/mockitointegration/NoByteCodeDependenciesTest.java and adjust it to ensure that we do not load the mockmakers.

E.g. load both `Mockito.class` and `Matchers.class` and assert that it does **not** load any of the mockmakers that we ship (`.without("org.mockito.internal.creation.bytebuddy", "org.mockito.internal.creation.proxy")` on the ` ClassLoaders.excludingClassLoader()`)

JojOatXGME replied:
I adjusted the implementation, but I have not yet implemented a test.

I think the test you have suggested might not be sufficient. At least the JVMs which I am using do not load classes used by methods before the method is called. This means the test you have suggested succeeds with my previous implementation, at least on the JVM I am using. I think I would actually have to create a mock with another mock maker, while the built-in mock makers are not accessible. Maybe I will look at this later this week.

TimvdLippe replied:
Hm good point. In that case, let's punt it for now.

TimvdLippe commented on src/main/java/org/mockito/internal/configuration/SpyAnnotationEngine.java:
```diff
@@ -83,6 +83,7 @@ public AutoCloseable process(Class<?> context, Object testInstance) {
     }
 
     private static Object spyInstance(Field field, Object instance) {
+        // TODO: Should we also add an option to @Spy?
```
Yes, but let's file a separate GitHub issue for that and link it here: https://github.com/mockito/mockito/issues/2740

JojOatXGME replied:
Ok. I have added the issue number to the comment.

TimvdLippe commented on src/main/java/org/mockito/plugins/MockMaker.java:
```diff
@@ -45,6 +46,15 @@
+ * you can specify it using {@link MockSettings#mockMaker(String)}.</p>
+ * <pre>
+ *     Object mock = Mockito.mock(Object.class, Mockito.withSettings()
+ *             .mockMaker(AwesomeMockMaker.class));
```
Nit: this should be updated to use the String. Let's include two examples here:

```java
 * <pre>
 *     // Use a mockmaker shipping with Mockito
 *     Object mock = Mockito.mock(Object.class, Mockito.withSettings()
 *             .mockMaker(MockMakers.INLINE));
 *     // Or load a mockmaker using a fully qualified class name
 *     Object mock = Mockito.mock(Object.class, Mockito.withSettings()
 *             .mockMaker("com.company.mockito.MyAwesomeMockMaker"));
 * </pre>
```

JojOatXGME replied:
Fixed.

TimvdLippe reviewed (changes requested):
Nice, yes this is looking a lot better. Very happy about this design choice, this looks natural for loading custom mockmakers as well. I like your solution of `MockMakers.java`, since these strings were kinda magic at first and now we have a proper dynamic API for them as well.

Some last minor changes, but the overall design LGTM now.

We should also include a nice mention in our main Javadoc. Can you add section 53 to `Mockito.java`? Something along the lines of:

```
 * <h3 id="53">53. <a class="meaningful_link" href="#mock_annotation_mockmaker" name="mock_annotation_mockmaker">
 *  New mockmaker attribute for @Mock annotation and <code>MockSettings.mockMaker(String)</code> methods (Since 4.6.0)</a></h3>
 *
 * You can now customize the mockmaker for a single mock, either using `@Mock` annotation mockmaker attribute or
 * using `MockSettings.mockMaker(String)`. This can be useful for migrating to a different mockmakers
 * (which we are planning to do in Mockito 5 with https://github.com/mockito/mockito/issues/2589).
 *
 * <pre class="code"><code class="java">
 *   &#064;Mock(mockMaker = MockMakers.INLINE)
 *   Foo mock;
 *   // using MockSettings.withSettings()
 *   Foo mock = Mockito.mock(Foo.class, withSettings().mockMaker(MockMakers.SUBCLASS));
 * </code></pre>
```

Also, make sure you add a link to section 53 at the top of the Javadoc section.

JojOatXGME commented on src/main/java/org/mockito/internal/configuration/plugins/DefaultMockitoPlugins.java:
```diff
@@ -59,7 +71,8 @@ public <T> T getDefaultPlugin(Class<T> pluginType) {
         return create(pluginType, className);
     }
 
-    String getDefaultPluginClass(String classOrAlias) {
+    public String getDefaultPluginClass(String classOrAlias) {
+        // TODO: #2701 We could make this method static?
         return DEFAULT_PLUGINS.get(classOrAlias);
     }
```
@TimvdLippe Are there any objections for making this method static?

JojOatXGME commented on src/main/java/org/mockito/internal/util/MockUtil.java:
```diff
@@ -18,22 +19,68 @@
+        }
+
+        Class<? extends MockMaker> type;
+        // TODO: #2701 I doubt that it is a good idea to use the context class loader here,
+        //  but PluginInitializer.loadImpl is using it as well,
+        //  but DefaultMockitoPlugins.create is explicitly not using it.
+        ClassLoader loader = Thread.currentThread().getContextClassLoader();
```
Using `Thread.currentThread().getContextClassLoader()` could cause problems with our built-in mock makers (see pre-existing comment in [`DefaultMockitoPlugins.create`](https://github.com/mockito/mockito/blob/530558ae3c721cb310f3410a66ff4061414cf836/src/main/java/org/mockito/internal/configuration/plugins/DefaultMockitoPlugins.java#L80-L83)). It might also cause a memory leak if someone is loading a lot of mock makers from temporary class loaders.

However, `PluginInitializer.loadImpl` is also using the context class loader, which is even more problematic in my opinion. The result of `PluginInitializer.loadImpl` is stored in static fields (i.e. `MockUtil.defaultMockMaker` and `Plugins.registry.mockMaker`), so Mockito considers the context class loader only when these static fields are first initialized. This probably means that the test which is executed first determines the mock maker which is used as the default, which seems quite arbitrary to me.

@TimvdLippe, do you know if there is any history about the usage of context class loaders with Mockito? Or do you have any opinion on that?

TimvdLippe replied:
No I am not ware of any history. Usually, if there are issues with ClassLoaders, we write specific integration tests and then we fix them. In this case, since I am not aware of any problems, I think it's fine to keep it as-is.

JojOatXGME reviewed (commented):
I made some adjustments. I want to the Javadoc later this week. There are also a few remaining TODOs I would like to discuss before merging this pull request.

JojOatXGME commented:
@TimvdLippe, I have finished my work. If you have no further suggestions, I would replace my `${NEXT_VERSION}` placeholder with `4.8.0` (or should it be `4.7.1`?) and rebase everything into one commit. Or what is the process?

**FYI:** I noticed that the following test is flaky, even on the `main` branch.

https://github.com/mockito/mockito/blob/530558ae3c721cb310f3410a66ff4061414cf836/src/test/java/org/mockito/internal/stubbing/defaultanswers/ReturnsDeepStubsConcurrentTest.java#L17-L24

It sometimes throws an `ArrayIndexOutOfBoundsException` in `InvocationContainerImpl.getStubbingsAscending`. The exception seems to be caused by a race condition. While this thread is reading `InvocationContainerImpl.stubbed`, another thread seems to add or remove items from the list.

TimvdLippe commented:
@JojOatXGME Great work! The next version will be 4.8.0, so let's use that in the documentation 😄 

Regarding the flaky test: if you have time to fix it, please go ahead. But for now, since we retry our tests, I think we can keep it as-is for now. Not ideal, I know...

JojOatXGME commented:
@TimvdLippe I replaced my version placeholder and rebased and squashed my commits. I completely forgot [your suggestion](https://github.com/mockito/mockito/pull/2701#pullrequestreview-1087903980) while writing the Javadoc, so if you have any final remarks or prefer another wording, I am happy to make adjustments. Otherwise, I think there is nothing left to be done for this PR.

TimvdLippe commented on src/test/java/org/mockitointegration/DeferMockMakersClassLoadingTest.java:
```diff
@@ -0,0 +1,96 @@
+import org.mockito.plugins.MockMaker;
+import org.mockitoutil.ClassLoaders;
+
+public class DeferMockMakersClassLoadingTest {
```
Nice test!

TimvdLippe reviewed (approved):
Woohoo! Great feature 😄 

Running the final checks on CI and then I will merge.

TimvdLippe commented:
Released as part of 4.8.0 🎉
